src/dmd/mars.d

```diff
@@ -2125,3 +2168,278 @@ private bool parseCommandLine(const ref Strings arguments, const size_t argc, re
+{
+    if (includeImports)
+    {
+        Identifiers empty;
+        if (includeImportedModuleCheck(ModuleComponentRange(
+            (m.md && m.md.packages) ? m.md.packages : &empty, m.ident, m.isPackageFile)))
+        {
+            if (global.params.verbose)
+                fprintf(global.stdmsg, "compileimport (%s)\n", m.srcfile.toChars);
+            compiledImports.push(m);
+            return true; // this import will be compiled
+        }
+    }
+    return false; // this import will not be compiled
+}
+
+// A range of component identifiers for a module
+private struct ModuleComponentRange
+{
+    Identifiers* packages;
+    Identifier name;
+    bool isPackageFile;
+    size_t index;
+    @property auto totalLength() const { return packages.dim + 1 + (isPackageFile ? 1 : 0); }
+
+    @property auto empty() { return index >= totalLength(); }
+    @property auto front() const
+    {
+        if (index < packages.dim)
+            return (*packages)[index];
+        if (index == packages.dim)
+            return name;
+        else
+            return Identifier.idPool("package");
+    }
+    void popFront() { index++; }
+}
+
+/*
+ * Determines if the given module should be included in the compilation.
+ * Returns:
+ *  True if the given module should be included in the compilation.
+ */
+private bool includeImportedModuleCheck(ModuleComponentRange components)
+    in { assert(includeImports); } body
+{
+    createMatchNodes();
+    size_t nodeIndex = 0;
+    while (nodeIndex < matchNodes.dim)
+    {
+        //printf("matcher ");printMatcher(nodeIndex);printf("\n");
+        auto info = matchNodes[nodeIndex++];
+        if (info.depth <= components.totalLength())
+        {
+            size_t nodeOffset = 0;
+            for (auto range = components;;range.popFront())
+            {
+                if (range.empty || nodeOffset >= info.depth)
+                {
+                    // MATCH
+                    //printf("matcher ");printMatcher(nodeIndex - 1);
+                    //printf(" MATCHES module '");components.print();printf("'\n");
+                    return !info.isExclude;
+                }
+                if (!range.front.equals(matchNodes[nodeIndex + nodeOffset].id))
+                {
+                    break;
+                }
+                nodeOffset++;
+            }
+        }
+        //printf("matcher ");printMatcher(nodeIndex-1);
+        //printf(" does not match module '");components.print();printf("'\n");
+        nodeIndex += info.depth;
+    }
+    assert(nodeIndex == matchNodes.dim, "code bug");
+    return includeByDefault;
+}
+
+// Matching module names is done with an array of matcher nodes.
+// The nodes are sorted by "component depth" from largest to smallest
+// so that the first match is always the longest (best) match.
+private struct MatcherNode
+{
+    union
+    {
+        struct
+        {
+            ushort depth;
+            bool isExclude;
```

Comment:
what does `isExclude` for?

Reply:
A switch for include patterns (`-i=foo.bar`) vs exclude patterns (`-i=-foo.bar`).  Check out the link to the documentation for a better explanation.

Reply:
Looking at the code, I guess you were probably more wondering what the `MatcherNode` is used for.  The full pattern-matching algorithm uses an array of matcher nodes.  This array of matcher nodes is a representation of all the patterns given to the compiler.  Each "pattern" has a `MatcherNode` that indicates how many components it has and whether it is an exclude pattern (via `isExclude`) followed by a number of "Identifier" `MatcherNode`'s which provide the identifiers for the pattern.

For example, the following "-i=-foo.bar -i=foo" would become:

Index | Node
---|---
0 | MatcherNode(depth=2, Yes.isExclude)
1 | MatcherNode(Id.foo)
2 | MatcherNode(Id.bar)
3 | MatcherNode(depth=1, No.isExclude)
1 | MatcherNode(Id.foo)

However I think you're main point is that this needs to be documented?


```